go/multipooler/connstate/connection_state.go

```diff
@@ -0,0 +1,391 @@
+}
+
+// =============================================================================
+// Settings - Session variables with Vitess-style bucket management
+// =============================================================================
+
+// globalSettingsCounter is used to assign unique bucket numbers to Settings.
+// This follows the Vitess pattern for distributing connections across pool stacks.
+var globalSettingsCounter atomic.Uint32
+
+// Settings contains session variables (SET commands) and a bucket number
+// for connection pool distribution.
+//
+// The bucket is assigned when the Settings is created and is used by the
+// connection pool to distribute connections with the same settings to the
+// same stack, enabling efficient connection reuse.
+//
+// IMPORTANT: Settings should be created via SettingsCache.GetOrCreate() to ensure
+// proper interning. When settings are interned, pointer equality can be used for
+// fast comparison instead of comparing the full Vars map.
+type Settings struct {
+	// Vars maps variable names to their values.
+	Vars map[string]string
+
+	// bucket is assigned from globalSettingsCounter when Settings is created.
+	// Used by connection pool for stack distribution.
+	bucket uint32
+}
+
+// NewSettings creates a new Settings with the given variables.
+// A unique bucket number is assigned from the global counter.
+//
+// NOTE: For connection pooling, prefer using SettingsCache.GetOrCreate() instead
+// to ensure settings are properly interned (same settings = same pointer).
+func NewSettings(vars map[string]string) *Settings {
+	return &Settings{
+		Vars:   vars,
+		bucket: globalSettingsCounter.Add(1),
+	}
+}
+
+// Bucket returns the bucket number for these settings.
+// This is used by the connection pool for stack distribution.
+func (s *Settings) Bucket() uint32 {
+	if s == nil {
+		return 0
+	}
+	return s.bucket
+}
+
+// ApplyQuery returns the SQL to apply these settings to a connection.
+func (s *Settings) ApplyQuery() string {
+	if s == nil || len(s.Vars) == 0 {
+		return ""
+	}
+
+	// Sort keys for deterministic output
+	keys := make([]string, 0, len(s.Vars))
+	for k := range s.Vars {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	// Build apply query
+	var b strings.Builder
+	for i, k := range keys {
+		if i > 0 {
+			b.WriteString("; ")
+		}
+		b.WriteString("SET SESSION ")
+		b.WriteString(k)
+		b.WriteString(" = '")
+		b.WriteString(s.Vars[k])
+		b.WriteString("'")
+	}
+	return b.String()
+}
+
+// ResetQuery returns the SQL to reset these settings on a connection.
+func (s *Settings) ResetQuery() string {
+	if s == nil || len(s.Vars) == 0 {
+		return ""
+	}
+	return "RESET ALL"
+}
+
+// IsEmpty returns true if there are no variables set.
+func (s *Settings) IsEmpty() bool {
+	if s == nil {
+		return true
+	}
+	return len(s.Vars) == 0
+}
+
+// Equals checks if two Settings are equal (same variables).
+// Note: bucket numbers are NOT compared since two Settings with the same
```

Comment:
I don't have strong feelings on this, but would it be unreasonable or awkward either rename this to something other than Equals() or split apart the settings storage from the bucket ID so this comment isn't necessary?

Feels a little unexpected to have an Equals() method that intentionally ignores fields

Reply:
That function was actually unused. Because if the Settings are created through the SettingsCache, then they will have the same pointer, and bucket number. Otherwise the bucket number is pretty much useless. Lol.